### PR TITLE
Add tooltip for longer card names

### PR DESCRIPTION
**Pull request by mrpau-eugene**
### Summary

Still need some improvements and would also like some suggestions if there would be better approach for showing the tooltip.

# Before:
screen shot 2018-03-01 at 3 16 07 pm

# After:
screen shot 2018-03-01 at 3 15 47 pm

There's a small problem after wrapping the `<router-link>` with a `<span>` (though I'm not quite sure if this is the best practice for adding tooltips) since it's causing a problem with the bottom margin for the content cards.

Submitting this PR for asking other suggestions and also whether to know if I'm on the right path 🙈 

### Reviewer guidance

…

### References

#2193 

----

### Contributor Checklist

- [ ] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [ ] Contributor has fully tested the PR manually
- [x] Screenshots of any front-end changes are in the PR description
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')

### Reviewer Checklist

- [ ] Automated test coverage is satisfactory
- [ ] Reviewer has fully tested the PR manually
- [ ] PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- [ ] External dependencies files were updated (`yarn` and `pip`)
- [ ] Documentation is updated
- [ ] Link to diff of internal dependency change is included
- [ ] CHANGELOG.rst is updated for high-level changes
- [ ] Contributor is in AUTHORS.rst

### CODE REVIEW

**Comment by indirectlylit** [comment]
Hi @mrpau-eugene 

Looks like a good start!

Two quick points:
* I'm not actually able to see the tooltip
* We should be able to show three lines of text. The old version (almost) did, but the new version does not:

| old | new |
| ---| --------|
| image | image |

I'll check out the code and experiment a little now.

File: kolibri/plugins/learn/assets/src/views/content-card/index.vue [hunk 1]
```diff
@@ -1,18 +1,23 @@
...
+
+    <router-link :to="link" class="card" :class="{ 'mobile-card': isMobile }">
+      <card-thumbnail
+        class="thumbnail"
+        :thumbnail="thumbnail"
+        :kind="kind"
+        :progress="progress"
+        :isMobile="isMobile"
+      />
+      <h3 class="text" dir="auto" v-shave>{{ title }}</h3>
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/views/content-card/index.vue]: let's attach the tooltip to just the title, rather than the whole card

File: kolibri/plugins/learn/assets/src/views/content-card/index.vue [hunk 2]
```diff
@@ -62,6 +77,16 @@
         default: false,
       },
     },
+    data() {
+      return {
+        hasTooltip: false,
+      };
+    },
+    updated() {
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/views/content-card/index.vue]: I believe that this only gets triggered when vue's updates the dom, but not when the dom is resized.

In order to handle resize events, we can probably leverage our "responsive element" mixin.

**Comment by indirectlylit** [comment]
It looks like the `ui-tooltip` component doesn't handle large text strings too well:

image

Instead we should probably just use the native tooltip like this:

image

File: kolibri/plugins/learn/assets/src/views/content-card/index.vue [hunk 3]
```diff
@@ -62,6 +77,16 @@
...
+      this.$nextTick(() => {
+        this.hasTooltip = this.$el.querySelector('h3 .js-shave') !== null;
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/views/content-card/index.vue]: This selector is a bit brittle - it will break if we change the template DOM structure.

Instead you can add a `ref` to the `h3` element and then reference it directly:

```javascript
 this.hasTooltip = Boolean(this.$refs.title.querySelector('.js-shave'));
```

File: kolibri/plugins/learn/assets/src/views/content-card/index.vue [hunk 4]
```diff
@@ -1,18 +1,23 @@
...
-      :thumbnail="thumbnail"
-      :kind="kind"
-      :progress="progress"
-      :isMobile="isMobile"
-    />
-
-    <h3 class="text" dir="auto">{{ title }}</h3>
-
-  </router-link>
+  <span ref="complete-title">
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/views/content-card/index.vue]: > There's a small problem after wrapping the `<router-link>` with a `<span>` (though I'm not quite sure if this is the best practice for adding tooltips) since it's causing a problem with the bottom margin for the content cards.

Why was this necessary? Why not add the `ref` directly to the `router-link` component?

**Comment by indirectlylit** [comment]
submitted a PR here with some possible changes corresponding to the comments I left above: https://github.com/mrpau/kolibri/pull/1


These are just options - recommend reading carefully and deciding if you like them.

**Comment by indirectlylit** [comment]
I'd also look carefully at the performance here. After some brief testing, I believe that the `vue-shave` directive is updating all of these DOM elements FAR too often based on global resize events: https://github.com/quantity-digital/vue-shave/blob/master/src/vue-shave.js

I'm thinking we should remove `vue-shave` and instead write our own wrapper around `shave` which leverages our `responsive-element` mixin to decide when to run the shavers, since we can more efficiently listening for element-specific resize events using that: https://github.com/learningequality/kolibri/blob/develop/kolibri/core/assets/src/mixins/responsive-element.js

**Comment by indirectlylit** [comment]
This isn't working yet, but I'd recommend creating a new `shaved-text` component that is similar to this:

https://gist.github.com/indirectlylit/dea91635bf71974334c3f02ee120f29a

then we can make it a core component and re-use it in other places.

File: kolibri/plugins/learn/assets/src/views/content-card/index.vue (hunk 2 above)
Comment (mrpau-eugene) [reply kolibri/plugins/learn/assets/src/views/content-card/index.vue]: > I believe that this only gets triggered when vue's updates the dom, but not when the dom is resized.

Yeah, I was looking for other alternatives in Vue's documentation but couldn't find an event that fires when the dom is resized.

File: kolibri/plugins/learn/assets/src/views/content-card/index.vue (hunk 3 above)
Comment (mrpau-eugene) [reply kolibri/plugins/learn/assets/src/views/content-card/index.vue]: Cool! I didn't know you can do something like this

File: kolibri/plugins/learn/assets/src/views/content-card/index.vue (hunk 4 above)
Comment (mrpau-eugene) [reply kolibri/plugins/learn/assets/src/views/content-card/index.vue]: I did something like this for my first attempt

```
  <router-link :to="link" class="card" :class="{ 'mobile-card': isMobile }" ref="complete-title">
    <card-thumbnail
      class="thumbnail"
      :thumbnail="thumbnail"
      :kind="kind"
      :progress="progress"
      :isMobile="isMobile"
    />
    <h3 class="text" dir="auto" v-shave>{{ title }}</h3>
    <ui-tooltip v-if="hasTooltip" trigger="complete-title">
      {{ title }}
    </ui-tooltip>
  </router-link>
```

But it was having a weird output so I thought I needed to wrap it inside an element
screen shot 2018-03-05 at 12 27 55 pm

**Comment by mrpau-eugene** [comment]
> I'm thinking we should remove vue-shave and instead write our own wrapper around shave which leverages our responsive-element mixin to decide when to run the shavers, since we can more efficiently listening for element-specific resize events using that

+1 for this!

> This isn't working yet, but I'd recommend creating a new shaved-text component that is similar to this

Will do this one. Should I place it somewhere in `kolibri/core/asserts/views` directory and make something similar like the `k-select`, `k-checkbox`, etc.. ?

**Comment by indirectlylit** [comment]
> Will do this one. Should I place it somewhere in kolibri/core/asserts/views directory and make something similar like the k-select, k-checkbox, etc.. ?

For now, just put it at `content-card/shaved-text.vue`. Once it's done we can move it to core.

**Comment by indirectlylit** [comment]
this looks great - nice job@
